app/controllers/topics_controller.rb

```diff
@@ -9,60 +12,52 @@ def new
 
   def create
     @topic = Topic.new(create_topic_params)
+
     if params[:add_heading]
-      @topic.topic_sections.build(position: next_position_in_list(@topic))
-      render :edit
-      return
+      add_heading(@topic)
+    else
+      respond_for_topic_publication save_draft(@topic), notice: "Topic has been created"
     end
-
-    publication = TopicPublisher.new(content_model: @topic)
-      .save_draft(TopicPresenter.new(@topic))
-
-    respond_for_topic_publication publication, notice: "Topic has been created"
   end
 
   def edit
-    @topic = Topic.find(params[:id])
   end
 
   def update
-    @topic = Topic.find(params[:id])
     @topic.assign_attributes(update_topic_params)
-    if params[:add_heading]
-      @topic.topic_sections.build(position: next_position_in_list(@topic))
-      render :edit
-      return
-    end
 
-    if params[:publish]
-      publish
+    if params[:add_heading]
+      add_heading(@topic)
+    elsif params[:publish]
+      respond_for_topic_publication publish(@topic), notice: "Topic has been published"
     else
-      save_draft
+      respond_for_topic_publication save_draft(@topic), notice: "Topic has been updated"
```

Comment:
I find it weird that we're using `respond_for_topic_publication` to handle the output of a draft being saved, maybe it needs to be renamed?